File: javascript/luhn/luhn.js
function buildAddIns(numberText, alternating) {
  var addends = [];
  for (var i = numberText.length - 1; i >= 0; i--) {
    var n = parseInt(numberText[i]);
    var r = alternating ? n * 2 : n;
    if (r > 9) {
      r -= 9;
    }
    addends.push(r);
    alternating = !alternating;
  }
  return addends.reverse();
}

var Luhn = function (number) {
  this.numberText = number.toString();
  this.checkDigit = parseInt(this.numberText[this.numberText.length - 1]);
  this.addends = buildAddIns(this.numberText, false);
  this.checksum = this.addends.reduce(function (a, b) {
    return a + b;
  });
  this.valid = this.checksum % 10 == 0;
};

Luhn.create = function (number) {
  var numberText = number.toString();
  var addEnds = buildAddIns(numberText, true);
  var checksum = addEnds.reduce(function (a, b) {
    return a + b;
  });
  if (checksum % 10 == 0) {
    numberText += '0';
  } else {
    var checksumText = checksum.toString();
    var lastNumber = parseInt(checksumText[checksumText.length - 1]);
    numberText += (10 - lastNumber).toString();
  }
  return parseInt(numberText);
};

module.exports = Luhn;

File: csharp/luhn/Luhn.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Luhn
{
  public static int[] GenerateAddEnds(string numberString, bool alternating)
  {
    var addEnds = new List<int>();    
    for (var i = numberString.Length - 1; i >= 0; i--)
    {
      if (alternating)
      {
        var addEnd = (int)char.GetNumericValue(numberString[i]) * 2;
        addEnds.Add(addEnd > 9 ? addEnd - 9 : addEnd);
      }
      else
      {
        addEnds.Add((int)char.GetNumericValue(numberString[i]));
      }
      alternating = !alternating;
    }
    addEnds.Reverse();
    return addEnds.ToArray();
  }

  public Luhn(int aNumber)
  {
    var numberString = aNumber.ToString();
    CheckDigit = (int)char.GetNumericValue(numberString[numberString.Length - 1]);
    
    Addends = GenerateAddEnds(numberString, false);
    Checksum = Addends.Sum();
    Valid = Checksum % 10 == 0;
  }

  public int CheckDigit { get; }
  public int[] Addends { get; }
  public int Checksum { get; }
  public bool Valid { get; }

  public static Int64 Create(int aNumber)
  {
    var numberString = aNumber.ToString();
    var addEnds = GenerateAddEnds(numberString, true);
    var sumEnds = addEnds.Sum();
    if (sumEnds % 10 == 0)
    {
      numberString += "0";
    }
    else
    {
      var sumString = sumEnds.ToString();
      var lastNumber = char.GetNumericValue(sumString[sumString.Length - 1]);
      numberString += (10 - lastNumber).ToString();
    }
    return Int64.Parse(numberString);
  }
}